src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/Target_X64/X64ReadyToRunHelperNode.cs

```diff
@@ -225,5 +225,99 @@ protected override void EmitCode(NodeFactory factory, ref X64Emitter encoder, bo
                     throw new NotImplementedException();
             }
         }
+
+        // emits code that results in ThreadStaticBase referenced in RAX.
+        // may trash volatile registers. (there are calls to the slow helper and possibly to platform's TLS support)
+        private static void EmitInlineTLSAccess(NodeFactory factory, ref X64Emitter encoder)
+        {
+            ISymbolNode getInlinedThreadStaticBaseSlow = factory.HelperEntrypoint(HelperEntrypoint.GetInlinedThreadStaticBaseSlow);
+            ISymbolNode tlsRoot = factory.TlsRoot;
+            // IsSingleFileCompilation is not enough to guarantee that we can use "Initial Executable" optimizations.
+            // we need a special compiler flag analogous to /GA. Just assume "false" for now.
+            // bool singleFileExe = factory.CompilationModuleGroup.IsSingleFileCompilation;
+            bool singleFileExe = false;
```

Comment:
If we already know isSingleFileExe is not what we want, can we call this what this is? (Same for ARM64)

Reply:
This will need to be changed anyways once the "optimize for application" is implemented. I assume we will be doing that.

But I can change this.  Would `isInitialExecutable` look better?

Reply:
I would name it after the expected optimization (canAssumeZeroOffset or whatever). `singleFileExe` is unlikely to be the thing we end up keying on because I don't expect we'd enable it whenever we build a single file exe (I don't see this option enabled by default for EXE projects in VC++ VS projects and I don't even see a UI option to enable it in VS settings in general besides typing `/GA` manually - both of those feel like red flags we'd need to understand).

Reply:
Not sure about Windows. Perhaps the difference on win-x64 is not big enough and building an ".exe" is not common enough. On win-arm64 there would be more difference, but perhaps they chose to be consistent with x64, or just did not think about the flag.

Both GCC and Clang assume single executable mode by default (`-fpie`). Dynamic TLS mode is basically a de-optimization that happens when building an .so (i.e. `-fpic` is used). 
The differences are more significant than in Windows. See: https://github.com/dotnet/runtime/blob/46b426aba5244de309109c2190c7d002db7eb331/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/Target_ARM64/ARM64ReadyToRunHelperNode.cs#L243

I do not think the flag should describe the difference in the codegen. because what happens is specific to OS and architecture. 
The flag should simply specify what user intends to produce - an executable or a dynamic library.